src/main/java/com/yahoo/bullet/rest/service/WebSocketService.java

```diff
@@ -0,0 +1,81 @@
+/*
+ *  Copyright 2018, Yahoo Inc.
+ *  Licensed under the terms of the Apache License, Version 2.0.
+ *  See the LICENSE file associated with the project for terms.
+ */
+package com.yahoo.bullet.rest.service;
+
+import com.yahoo.bullet.pubsub.Metadata;
+import com.yahoo.bullet.rest.model.WebSocketResponse;
+import com.yahoo.bullet.rest.query.WebSocketQueryHandler;
+import lombok.AccessLevel;
+import lombok.Getter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.messaging.simp.SimpMessageHeaderAccessor;
+import org.springframework.messaging.simp.SimpMessagingTemplate;
+import org.springframework.stereotype.Service;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Service
+public class WebSocketService {
+    private static final String DESTINATION =  "/response/private";
+    @Autowired
+    private SimpMessagingTemplate simpMessagingTemplate;
+
+    @Autowired
+    private QueryService queryService;
+
+    // Exposed for testing only.
+    @Getter(AccessLevel.PACKAGE)
+    private Map<String, String> sessionIDMap = new ConcurrentHashMap<>();
+
+    /**
+     * Send a KILL signal to the backend.
+     *
+     * @param sessionID The session id to represent the client.
+     * @param queryID The query id of the query to be killed or null to kill the query associated with the session.
+     */
+    public void sendKillSignal(String sessionID, String queryID) {
+        if (sessionIDMap.containsKey(sessionID)) {
+            if (queryID == null || queryID.equals(sessionIDMap.get(sessionID))) {
+                queryService.submitSignal(sessionIDMap.get(sessionID), Metadata.Signal.KILL);
+                removeSessionID(sessionID);
+            }
+        }
+    }
+
+    /**
+     * Remove a session id from the session id map.
+     *
+     * @param sessionID The session id to be removed.
+     */
+    public void removeSessionID(String sessionID) {
+        sessionIDMap.remove(sessionID);
+    }
+
+    /**
+     * Submit a query by {@link QueryService}.
+     *
+     * @param queryID The query ID to register request with.
+     * @param sessionID The session id to represent the client.
+     * @param queryString The String version of the query.
+     */
+    public void submitQuery(String queryID, String sessionID, String queryString) {
+        sessionIDMap.put(sessionID, queryID);
+        WebSocketQueryHandler queryHandler = new WebSocketQueryHandler(this, sessionID);
```

Comment:
If you do the acknowledge in the constructor(), then you should move this after the queryService submit

Reply:
If we want submit first then and acknowledge, I can not call acknowledge in constructor() since submit needs queryhandler.  So I call queryHandler.acknowledge() explicitly after submit.